x/virtualgroup/keeper/msg_server.go

```diff
@@ -0,0 +1,386 @@
+		Id: req.GlobalVirtualGroupId,
+	}); err != nil {
+		return nil, err
+	}
+	return &types.MsgDeleteGlobalVirtualGroupResponse{}, nil
+}
+
+func (k msgServer) Deposit(goCtx context.Context, req *types.MsgDeposit) (*types.MsgDepositResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	funcAcc := sdk.MustAccAddressFromHex(req.FundingAddress)
+
+	sp, found := k.spKeeper.GetStorageProviderByFundingAddr(ctx, funcAcc)
+	if !found {
+		return nil, sptypes.ErrStorageProviderNotFound
+	}
+
+	gvg, found := k.GetGVG(ctx, req.GlobalVirtualGroupId)
+	if !found {
+		return nil, types.ErrGVGNotExist
+	}
+
+	depositDenom := k.DepositDenomForGVG(ctx)
+	if depositDenom != req.Deposit.GetDenom() {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidDenom, "invalid coin denomination: got %s, expected %s", req.Deposit.Denom, depositDenom)
+	}
+
+	// deposit the deposit token to module account.
+	coins := sdk.NewCoins(sdk.NewCoin(depositDenom, req.Deposit.Amount))
+	if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, sp.GetFundingAccAddress(), types.ModuleName, coins); err != nil {
+		return nil, err
+	}
+
+	gvg.TotalDeposit = gvg.TotalDeposit.Add(req.Deposit.Amount)
+	k.SetGVG(ctx, gvg)
+
+	if err := ctx.EventManager().EmitTypedEvents(&types.EventUpdateGlobalVirtualGroup{
+		Id:           req.GlobalVirtualGroupId,
+		StoreSize:    gvg.StoredSize,
+		TotalDeposit: gvg.TotalDeposit,
+	}); err != nil {
+		return nil, err
+	}
+	return &types.MsgDepositResponse{}, nil
+}
+
+func (k msgServer) Withdraw(goCtx context.Context, req *types.MsgWithdraw) (*types.MsgWithdrawResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	funcAcc := sdk.MustAccAddressFromHex(req.FundingAddress)
+
+	sp, found := k.spKeeper.GetStorageProviderByFundingAddr(ctx, funcAcc)
+	if !found {
+		return nil, sptypes.ErrStorageProviderNotFound
+	}
+
+	gvg, found := k.GetGVG(ctx, req.GlobalVirtualGroupId)
+	if !found {
+		return nil, types.ErrGVGNotExist
+	}
+
+	depositDenom := k.DepositDenomForGVG(ctx)
+	if req.Withdraw.Denom != depositDenom {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidDenom, "invalid coin denomination: got %s, expected %s", req.Withdraw.Denom, k.DepositDenomForGVG(ctx))
+	}
+
+	var withdrawTokens math.Int
+
+	availableTokens := k.GetAvailableStakingTokens(ctx, gvg)
+	if req.Withdraw.Amount.IsZero() {
+		withdrawTokens = availableTokens
+	} else {
+		if availableTokens.LT(req.Withdraw.Amount) {
+			return nil, types.ErrWithdrawAmountTooLarge
+		}
+		withdrawTokens = req.Withdraw.Amount
+	}
+
+	// withdraw the deposit token from module account to funding account.
+	coins := sdk.NewCoins(sdk.NewCoin(depositDenom, withdrawTokens))
+	if err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, sp.GetFundingAccAddress(), coins); err != nil {
+		return nil, err
+	}
+
+	if err := ctx.EventManager().EmitTypedEvents(&types.EventUpdateGlobalVirtualGroup{
+		Id:           req.GlobalVirtualGroupId,
+		StoreSize:    gvg.StoredSize,
+		TotalDeposit: gvg.TotalDeposit,
```

Comment:
Can we change `gvg.TotalDeposit` to negative value? Or use another event. Otherwise, it is confused.

Reply:
Can't be negative value. I'll change it to `math.Uint`